forecasting/src/autogluon/forecasting/trainer/abstract_trainer.py

```diff
@@ -0,0 +1,668 @@
+        save_pkl.save(path=path, object=data, verbose=verbose)
+
+    def save_val_data(self, data: TimeSeriesDataFrame, verbose: bool = True) -> None:
+        path = self.path_data + "val.pkl"
+        save_pkl.save(path=path, object=data, verbose=verbose)
+
+    def load_train_data(self) -> TimeSeriesDataFrame:
+        path = self.path_data + "train.pkl"
+        return load_pkl.load(path=path)
+
+    def load_val_data(self) -> TimeSeriesDataFrame:
+        path = self.path_data + "val.pkl"
+        return load_pkl.load(path=path)
+
+    def load_data(self) -> Tuple[TimeSeriesDataFrame, TimeSeriesDataFrame]:
+        train_data = self.load_train_data()
+        val_data = self.load_val_data()
+        return train_data, val_data
+
+    def _add_model(self, model: AbstractForecastingModel):
+        # TODO: also register predict time
+        self.model_info[model.name] = {}
+        self.model_info[model.name]["path"] = model.path
+        self.model_info[model.name]["type"] = type(model)
+        self.model_info[model.name]["fit_time"] = model.fit_time
+        self.model_info[model.name]["val_score"] = model.val_score
+
+    def _train_single(
+        self,
+        train_data: TimeSeriesDataFrame,
+        model: AbstractForecastingModel,
+        val_data: Optional[TimeSeriesDataFrame] = None,
+        time_limit: Optional[float] = None,
+    ) -> AbstractForecastingModel:
+        """Train the single model and return the model object that was fitted. This method
+        does not save the resulting model."""
+        model.fit(train_data=train_data, val_data=val_data, time_limit=time_limit)
+        return model
+
+    def _train_single_full(
+        self,
+        train_data: TimeSeriesDataFrame,
+        model: AbstractForecastingModel,
+        val_data: Optional[TimeSeriesDataFrame] = None,
+        hyperparameter_tune: bool = False,
+        time_limit: Optional[float] = None,
+    ) -> List[str]:
+        if hyperparameter_tune:
+            scheduler_cls, scheduler_options = scheduler_factory("auto")
+            scheduler_options["num_trials"] = 10 if time_limit is None else None
+            hpo_models, hpo_model_performances, hpo_results = model.hyperparameter_tune(
+                train_data=train_data,
+                val_data=val_data,
+                scheduler_options=(scheduler_cls, scheduler_options),
+                time_limit=time_limit,
+            )
+            hpo_results.pop("search_strategy", None)
+            self.hpo_results[model.name] = hpo_results
+            model_names_trained = []
+
+            # add each of the trained HPO configurations to the trained models
+            for model_hpo_name, model_path in hpo_models.items():
+                model_hpo = self.load_model(
+                    model_hpo_name, path=model_path, model_type=type(model)
+                )
+                self._add_model(model_hpo)
+                model_names_trained.append(model_hpo.name)
```

Comment:
Just a note: Code for HPO is pretty hacky in AbstractTrainer in Tabular at the moment, probably deserves a refactor at some point so it isn't crammed into `_train_single_full` but instead has its own method. If this PR can lead the charge here by making a separate method for this in forecast so its tidier, I think it would be great.